### PR TITLE
intersphinx: Allow strict prefix matching

### LINKED ISSUE #2068: Add an option to not use intersphinx references as a fallback
That intersphinx references are used as fallbacks if a reference is not available in a project recently caused us a problem.

We have a site (`latest`) which is the latest build of `master`.

We use intersphinx as follows, so we can link to the latest documentation in some sections:

``` python
intersphinx_mapping = {
    'latest': ('http://doc-dev.clusterhq.com/', None),
}
```

``` rst
Make sure to follow the :ref:`latest documentation <latest:release-process>` when doing a release.
```

Our docs included a label (`.. foo:`) so this reference was available on `latest`.
This label was removed on a branch, along with one of the two references to this label.
In this case we would expect that building the branch would fail, as there is a reference to a label which doesn't exist.
Unexpectedly, the branch built successfully, because the label was found in `latest` as a fall back when it was not found locally. The branch was merged, and then later, when `latest` changed (because `master` was built again) builds stopped working because there was a reference to a non-existent label.

It would be good to have an option to not fall back, maybe something like `nitpicky`.

### PR DESCRIPTION
**Pull request by nijel**
Subject: Add optional strict prefix matching to intersphinx

### Feature or Bugfix

- Feature

### Purpose
- allows better control on what links go to the external documentation

### Detail
Strictly using prefix for intersphinx links gives a better control on
external links - only links explicitely declared will point to the
external documentation.

Fixes #2068

### Relates
- https://github.com/sphinx-doc/sphinx/issues/2068

### CODE REVIEW

File: doc/usage/extensions/intersphinx.rst [hunk 1]
```diff
@@ -148,6 +148,20 @@ linking:
       exception is raised if the server has not issued a response for timeout
       seconds.
 
+.. confval:: intersphinx_strict_prefix
+
+   .. versionadded:: 3.6.0
```
Comment (tk0miya) [inline doc/usage/extensions/intersphinx.rst]: The next release is 4.0. So this should be changed.

File: tests/test_ext_intersphinx.py [hunk 2]
```diff
@@ -173,6 +181,7 @@ def test_missing_reference_pydomain(tempdir, app, status, warning):
         'https://docs.python.org/': inv_file,
     }
     app.config.intersphinx_cache_limit = 0
+    app.config.intersphinx_strict_prefix = False
```
Comment (tk0miya) [inline tests/test_ext_intersphinx.py:184]: Is this initialization needed? I think the configuration is initialized as `False` by default. So this assignment is not needed. I understand setting it up explicitly in test_missing_reference. But I don't understand why are they added in other cases.

**Review by tk0miya — COMMENTED** [review]
LGTM with nits.

File: tests/test_ext_intersphinx.py (hunk 2 above)
Comment (nijel) [reply tests/test_ext_intersphinx.py:184]: I didn't work without that. Unfortunately, I have no clue how this works in the tests...

**Comment by jakobandersen** [comment]
Hmm, I'm not sure this is a descriptive name for the feature. How about something like ``intersphinx_resolve_only_explicit``?
But even then, this doesn't seem like the right way to go. As argued in #8418, it seems strange to me to ask a domain to resolve a reference that you don't actually want it to resolve, just so intersphinx can pick it up.
I suggest a more direct apporach with an intersphinx role for requesting an intersphinx resolution, and then more fine-grained config vars to disable the missing-reference event handler.

**Comment by nijel** [comment]
As for name, I'm open to change it to anything. Changing intersphinx massively is something I can't do.

Having a dedicated role for intersphinx as you describe https://github.com/sphinx-doc/sphinx/pull/8418#issuecomment-758838589 and having a way to disable handling of missing references would work for us as well. The drawback of this approach might be that it would require new Sphinx to build, while changing way the lookups works as done in this PR would keep the documentation compatible with current Sphinx as well (only the strict lookup would not be performed, what is fine as long as we do the validation in CI using compatible version).

I want to explicitly define where interphinx should be used and let Sphinx fail on missing references instead of silently picking up a reference from random linked documentation which happens to have a target which we forgot to define.

**Comment by phlax** [comment]
it would be incredibly useful to have this feature added.

without it intersphinx is basically unusable for versioning

**Comment by phlax** [comment]
@nijel do you still need this ? Any chance of rebasing/updating to resolve conflicts ?

if not, i could possibly take on this PR as it would be helpful to us to have this landed upstream

**Comment by nijel** [comment]
I've just rebased it...

**Comment by jakobandersen** [comment]
In connection with some of the other recent intersphinx discussions I think a single config var is too limiting. For example, one may like to disable intersphinx lookup for roles like ``:ref:`` and ``:doc:``, but keep it for all xref generated by source code constructs, e.g., all directives in the Python, C, C++, Javascript, RST, etc. domains.
Perhaps it should be a list of disallowed domains, which defaults to ``[]`` by in ``sphinx-quickstart`` gets set to ``['std']``? I guess this is the domain that gives the most surprising lookups?

**Comment by phlax** [comment]
> disable intersphinx lookup for roles like :ref: and :doc:, but keep it for all xref generated by source code constructs...a list of disallowed domains, which defaults to []

this sounds useful but my feeling is that it adds 2 layers of complexity

just being able to disable the lookup up for non explicitly tagged links is incredibly useful if you want to use intersphinx with different versions of the same docs (we are using @nijel 's patch and its working v well)

how about we land this and then i would potentially be up for contributing further in terms of adding more fine-grained controls over intersphinx lookups

**Comment by jakobandersen** [comment]
See #9459 for an implementation of https://github.com/sphinx-doc/sphinx/pull/8981#issuecomment-879815795.

**Comment by jakobandersen** [comment]
Replaced by https://github.com/sphinx-doc/sphinx/pull/9459, specifically by setting ``intersphinx_disabled_reftypes = ['*']``. Though ``intersphinx_disabled_reftypes = ['std:doc']`` should be enough to fix most silent resolution problems.

**Pull request closed without merging**